File: java-code/src/main/java/refactoring/conditional_logic/replace_conditional_with_polymorphism/after/Bird.java
package refactoring.conditional_logic.replace_conditional_with_polymorphism.after;

import lombok.AllArgsConstructor;
import lombok.Getter;

@Getter
@AllArgsConstructor
//STEP 4: Make Bird class abstract to escalate type specific behaviors to the subclasses
public abstract class Bird {
    private String name;
    private String type;
    private int numberOfCoconuts;
    private int voltage;
    private boolean isNailed;

    //STEP 1: Move bird specific methods to the class
    //STEP 2: Create subclasses for each kind of bird
    //STEP 9 : After refactoring BirdFunctions method, we dont need this method anymore
/*    public String plumage(Bird bird) {
        //STEP 6: call the method of subclasses
        return createBird(bird).plumage();
    }*/

    //STEP 10 : After refactoring BirdFunctions method, we dont need this method anymore
/*    public int airSpeedVelocity(Bird bird) {
        //STEP 6: call the method of subclasses
        return createBird(bird).airSpeedVelocity();
    }*/


    //STEP3: Create factory method for bird creation
    public Bird createBird(Bird bird) {
    return switch (bird.getType()) {
      case "EuropeanSwallow" -> new EuropeanSwallow(bird);
      case "AfricanSwallow" -> new AfricanSwallow(bird);
      case "NorwegianBlueParrot" -> new NorwegianParrot(bird);
      default -> new DefaultBird(bird);
    };
    }

    //STEP 5: Define behaviors to be implemented by the subclasses
    public abstract String plumage();
    public abstract int airSpeedVelocity();

}

File: go-code/conditional_logic/replace_conditional_with_polymorphism/example1/after/bird.go
package after

type BirdAttributes struct {
	Name             string
	Type             string
	NumberOfCoconuts int
	Voltage          int
	IsNailed         bool
}

// STEP 1 : Create Bird specific behaviors
// STEP 3: Create subclasses
type Bird interface {
	GetBirdAttributes() BirdAttributes
	Plumage() string
	AirSpeedVelocity() int
}

// STEP 2 : Refactor the functions to fix compilation errors
func plumages(birds []Bird) map[string]string {
	result := make(map[string]string)
	for _, b := range birds {
		// Step 5: Call factory directly from caller function
		result[b.GetBirdAttributes().Name] = CreateBird(b.GetBirdAttributes()).Plumage()
	}
	return result
}

// STEP 2.1 : Refactor the methods to fix compilation errors
func speeds(birds []Bird) map[string]int {
	result := make(map[string]int)
	for _, b := range birds {
		// Step 5.1: Call factory directly from caller function
		result[b.GetBirdAttributes().Name] = CreateBird(b.GetBirdAttributes()).AirSpeedVelocity()
	}
	return result
}

// STEP 2.2 : Refactor the methods to fix compilation errors
// Step 4: Decouple function from bird types
// Step 5: You dont need this function with bird input argument anymore
/*func plumage(bird Bird) string {
	return CreateBird(bird.GetBirdAttributes()).Plumage()
}*/

// STEP 2.3 : Refactor the methods to fix compilation errors
// Step 4.1: Decouple function from bird types
// Step 5.1: You dont need this function with bird input argument anymore
/*func airSpeedVelocity(bird Bird) int {
	return CreateBird(bird.GetBirdAttributes()).AirSpeedVelocity()
}*/
